data/abi/abi_value.go

```diff
@@ -0,0 +1,304 @@
+		value:   value,
+	}
+}
+
+// MakeByte takes a byte and returns an ABI Byte type value.
+func MakeByte(value byte) Value {
+	return Value{
+		ABIType: MakeByteType(),
+		value:   value,
+	}
+}
+
+// MakeAddress takes an [32]byte array and returns an ABI Address type value.
+func MakeAddress(value [32]byte) Value {
+	return Value{
+		ABIType: MakeAddressType(),
+		value:   value,
+	}
+}
+
+// MakeDynamicArray takes an array of ABI value (can be empty) and element type,
+// returns an ABI dynamic length array value.
+func MakeDynamicArray(values []Value, elemType Type) (Value, error) {
+	if len(values) >= (1 << 16) {
+		return Value{}, fmt.Errorf("dynamic array make error: pass in argument number larger than 2^16")
+	}
+	for i := 0; i < len(values); i++ {
+		if !values[i].ABIType.Equal(elemType) {
+			return Value{}, fmt.Errorf("type mismatch: %s and %s",
+				values[i].ABIType.String(), elemType.String())
+		}
+	}
+	return Value{
+		ABIType: MakeDynamicArrayType(elemType),
+		value:   values,
+	}, nil
+}
+
+// MakeStaticArray takes an array of ABI value and returns an ABI static length array value.
+func MakeStaticArray(values []Value) (Value, error) {
+	if len(values) >= (1 << 16) {
+		return Value{}, fmt.Errorf("static array make error: pass in argument number larger than 2^16")
+	} else if len(values) == 0 {
+		return Value{}, fmt.Errorf("static array make error: 0 argument passed in")
+	}
+	for i := 0; i < len(values); i++ {
+		if !values[i].ABIType.Equal(values[0].ABIType) {
+			return Value{}, fmt.Errorf("type mismatch: %s and %s",
+				values[i].ABIType.String(), values[0].ABIType.String())
+		}
+	}
+	return Value{
+		ABIType: MakeStaticArrayType(values[0].ABIType, uint16(len(values))),
+		value:   values,
+	}, nil
+}
+
+// MakeTuple takes an array of ABI values and returns an ABI tuple value.
+func MakeTuple(values []Value) (Value, error) {
+	if len(values) >= (1 << 16) {
+		return Value{}, fmt.Errorf("tuple make error: pass in argument number larger than 2^16")
+	}
+	tupleType := make([]Type, len(values))
+	for i := 0; i < len(values); i++ {
+		tupleType[i] = values[i].ABIType
+	}
+
+	castedTupleType, err := MakeTupleType(tupleType)
+	if err != nil {
+		return Value{}, err
+	}
+
+	return Value{
+		ABIType: castedTupleType,
+		value:   values,
+	}, nil
+}
+
+// MakeBool takes a boolean value and returns an ABI bool value.
+func MakeBool(value bool) Value {
+	return Value{
+		ABIType: MakeBoolType(),
+		value:   value,
+	}
+}
+
+// GetUint8 tries to retreve an uint8 from an ABI Value.
+func (v Value) GetUint8() (uint8, error) {
+	if v.ABIType.abiTypeID != Uint || v.ABIType.bitSize > 8 {
+		return 0, fmt.Errorf("value type unmatch or bitSize too large")
+	}
+	bigIntForm, err := v.GetUint()
+	if err != nil {
+		return 0, err
+	}
+	return uint8(bigIntForm.Uint64()), nil
+}
+
+// GetUint16 tries to retrieve an uint16 from an ABI Value.
+func (v Value) GetUint16() (uint16, error) {
+	if v.ABIType.abiTypeID != Uint || v.ABIType.bitSize > 16 {
```

Comment:
maybe all these uint size checks should be a small helper, checkUint(t Type, bitSize int)